src/charm.py

```diff
@@ -340,6 +370,79 @@ def _basic_auth_user(self) -> Optional[str]:
         """
         return cast(Optional[str], self.config.get("basic_auth_user", None))
 
+    def _validate_annotation_key(self, key: str) -> bool:
+        """Validate the annotation key."""
+        if len(key) > 253:
+            logger.error(f"Invalid annotation key: '{key}'. Key length exceeds 253 characters.")
+            return False
+
+        if not is_qualified_name(key.lower()):
+            logger.error(
+                f"Invalid annotation key: '{key}'. Must follow Kubernetes annotation syntax."
+            )
+            return False
+
+        if key.startswith(("kubernetes.io/", "k8s.io/")):
+            logger.error(f"Invalid annotation: Key '{key}' uses a reserved prefix.")
+            return False
+
+        return True
+
+    def _validate_annotation_value(self, value: str) -> bool:
+        """Validate the annotation value."""
+        if not ANNOTATION_VALUE_PATTERN.match(value):
+            logger.error(
+                f"Invalid annotation value: '{value}'. Must follow Kubernetes annotation syntax."
+            )
+            return False
+
+        return True
+
+    def _parse_annotations(self, annotations: Optional[str]) -> Optional[Dict[str, str]]:
+        """Parse and validate annotations from a string.
+
+        logic is based on Kubernetes annotation validation as described here:
+        https://github.com/kubernetes/apimachinery/blob/v0.31.3/pkg/api/validation/objectmeta.go#L44
+        """
+        if not annotations:
+            return None
+
+        annotations = annotations.strip().rstrip(",")  # Trim spaces and trailing commas
+
+        try:
+            parsed_annotations = {
+                key.strip(): value.strip()
+                for key, value in (pair.split("=", 1) for pair in annotations.split(",") if pair)
+            }
+        except ValueError:
+            logger.error(
+                "Invalid format for 'loadbalancer_annotations'. "
+                "Expected format: key1=value1,key2=value2."
+            )
+            return None
+
+        # Validate each key-value pair
+        for key, value in parsed_annotations.items():
+            if not self._validate_annotation_key(key) or not self._validate_annotation_value(
+                value
+            ):
+                return None
```

Comment:
extra context: my comment in `config.yaml` about whether to fail open or closed

imo this failure behaviour is too quiet.  Misconfiguring the LoadBalancer feels like something that should surface as a charm status.  Almost certainly if the user has provided annotations and we discard them, the LoadBalancer is not operating as they'd expect from a charm that says it is healthy.

`_validate_annotation_*()` do log errors (`logger.error(msg)`), but that is easily missed.

Reply:
Probably the reason we're discussing this though is because, like in #420, the current architecture of this charm makes handling an error that should block the whole charm really hard.  Ideally imo, this sort of misconfiguration would sit as a BlockedStatus until it was resolved (although maybe some parts of the charm would still function separately)

Given that we cannot easily achieve that atm, I think the LoadBalancer needs to break more obviously.  Maybe if we cannot validate the annotations, we should change the LoadBalancer service_type back to ClusterIP?